src/arithmetic/arithmetic_expression.c

```diff
@@ -1178,6 +1149,169 @@ SIValue AR_NE(SIValue *argv, int argc) {
+	}
+	return array;
+}
+
+// returns a value in a specific index in an array
+// valid index range is [-arrayLen, arrayLen)
+// invalid index will return null
+// "RETURN [1, 2, 3][0]" will yeild 1
+SIValue AR_SUBSCRIPT(SIValue *argv, int argc) {
+	assert(argc == 2 && argv[0].type == T_ARRAY && argv[1].type == T_INT64);
+	SIValue list = argv[0];
+	int32_t index = (int32_t)argv[1].longval;
+	uint32_t arrayLen = SIArray_Length(list);
+	// given a negativ index, the accses is calculated as arrayLen+index
+	uint32_t absIndex = abs(index);
+	// index range can be [-arrayLen, arrayLen) (lower bound inclusive, upper exclusive)
+	// this is because 0 = arrayLen+(-arrayLen)
+	if((index < 0 && absIndex > arrayLen) || (index > 0 && absIndex >= arrayLen)) return SI_NullVal();
+	SIValue res = index >= 0 ? SIArray_Get(list, index) : SIArray_Get(list,
+																	  arrayLen - absIndex);
+	// clone is in case for nested heap allocated values returned from the array
+	return SI_CloneValue(res);
+}
+
+// return a sub array from an array given a range of indices
+// valid indices ragne is [-arrayLen, arrayLen)
+// if range start value is bigger then range end value an empty list will be returnd
+// if indices are still integers but not in the valid range, only values within the valid range
+// will be returned
+// if one of the indices is null, null will be returnd
+// "RETURN [1, 2, 3][0..1]" will yield [1, 2]
+SIValue AR_SLICE(SIValue *argv, int argc) {
+	assert(argc == 3 && argv[0].type == T_ARRAY);
+	if(argv[0].type == T_NULL || argv[1].type == T_NULL || argv[2].type == T_NULL) return SI_NullVal();
+	assert(argv[1].type == T_INT64 && argv[2].type == T_INT64);
+	SIValue array = argv[0];
+
+	// get array length
+	uint32_t arrayLen = SIArray_Length(array);
+
+	// get start and end index
+	SIValue start = argv[1];
+	int32_t startIndex = (int32_t)start.longval;
+	SIValue end = argv[2];
+	int32_t endIndex = (int32_t)end.longval;
+
+	uint32_t absStartIndex = abs(startIndex);
+	// if negative index, calculate offset from end
+	startIndex = startIndex >= 0 ? startIndex : arrayLen - absStartIndex;
+	// if offset from the end is out of bound, start at 0
+	startIndex = startIndex < 0 ? 0 : startIndex;
+
+	uint32_t absEndIndex = abs(endIndex);
+	// if negative index, calculate offset from end
+	endIndex = endIndex >= 0 ? endIndex : arrayLen - absEndIndex;
+	// if index out of bound, end at arrayLen
+	endIndex = ((int32_t)arrayLen) < endIndex ? arrayLen : endIndex;
+	// cant go in reverse
+	if(endIndex <= startIndex) {
+		return SI_EmptyArray();
+	}
+
+	SIValue subArray = SI_Array(endIndex - startIndex);
+	for(uint i = startIndex; i < endIndex; i++) {
+		SIArray_Append(&subArray, SIArray_Get(array, i));
+	}
+	return subArray;
+}
+
+// create a new list of integers in the range of [start, end]. If an interval was given
+// the interval between two consecutive list members will be this interval.
+// If interval was not suppllied, it will be default as 1
```

Comment:
"Interval" refers to the range between beginning and end, the optional third argument here is usually referred to as "step".

Reply:
fixd